// @render-props/events/cjs/Events.js
'use strict'

var _interopRequireDefault = require('@babel/runtime/helpers/interopRequireDefault')

exports.__esModule = true
exports.default = void 0

var _react = _interopRequireDefault(require('react'))

var _propTypes = _interopRequireDefault(require('prop-types'))

class Events extends _react.default.Component {
  constructor(props) {
    super(props)
    this.listeners = []

    this.addEvent = (el, name, fn) => {
      this.listeners.push([el, name, fn])
      el.addEventListener(name, fn)
    }

    this.removeEvent = (el, name, fn) => {
      el.removeEventListener(name, fn)

      for (let x = 0; x < this.listeners.length; x++) {
        const event = this.listeners[x]
        const [el_, name_, fn_] = event

        if (el === el_ && name === name_ && fn === fn_) {
          this.listeners.splice(x, 1)
        }
      }
    }

    this.removeAllEvents = el => {
      for (let x = 0; x < this.listeners.length; x++) {
        const [el_, name, fn] = this.listeners[x]

        if (!el || el_ === el) {
          el_.removeEventListener(name, fn)
          this.listeners.splice(x, 1)
        }
      }
    }

    this.renderProps = {
      addEvent: this.addEvent,
      removeEvent: this.removeEvent,
      removeAllEvents: this.removeAllEvents,
    }
  }

  componentWillUnmount() {
    this.removeAllEvents()
  }

  render() {
    return this.props.children(this.renderProps)
  }
}

exports.default = Events
Events.propTypes = {
  children: _propTypes.default.func.isRequired,
}
