// @render-props/image-props/cjs/ImageProps.js
'use strict'

var _interopRequireDefault = require('@babel/runtime/helpers/interopRequireDefault')

exports.__esModule = true
exports.default = void 0

var _extends2 = _interopRequireDefault(
  require('@babel/runtime/helpers/extends')
)

var _react = _interopRequireDefault(require('react'))

var _propTypes = _interopRequireDefault(require('prop-types'))

var _viewport = require('@render-props/viewport')

var _utils = require('./utils')

class ImageProps extends _react.default.Component {
  constructor(props) {
    super(props)

    this.imageRef = e => {
      if (
        e !== this.element ||
        (e && this.element && e.src !== this.element.src)
      ) {
        this.element = e
        this.setStats()
      }
    }

    this.element = null
    this.imageLoader = null
    this.state = {
      orientation: 'square',
      width: 0,
      height: 0,
      naturalWidth: 0,
      naturalHeight: 0,
      complete: false,
    }
    this.statContext = {
      imageRef: this.imageRef,
    }
  }

  componentWillUnmount() {
    if (this.imageLoader !== null) {
      this.imageLoader.cancel()
    }
  }

  setStats() {
    if (this.element !== null) {
      this.imageLoader = (0, _utils.loadImage)(this.element)
      this.imageLoader.then(({target}) => {
        const {width, height, naturalWidth, naturalHeight} = target
        this.setState({
          orientation: (0, _viewport.getOrientation)({
            width: naturalWidth,
            height: naturalHeight,
          }),
          width,
          height,
          naturalWidth,
          naturalHeight,
          complete: true,
        })
        this.imageLoader = null
      })
    }
  }

  render() {
    return this.props.children(
      (0, _extends2.default)(this.statContext, this.state)
    )
  }
}

exports.default = ImageProps
ImageProps.propTypes = {
  children: _propTypes.default.func.isRequired,
}
