// @render-props/utils/cjs/index.js
'use strict'

exports.__esModule = true
exports.toKebabCaseTrimmed = exports.clearRequestTimeout = exports.clearRequestInterval = exports.cancelAnimationFrame = void 0

var _bound2 = _interopRequireDefault(require('./bound'))

exports.bound = _bound2.default

var _callIfExists2 = _interopRequireDefault(require('./callIfExists'))

exports.callIfExists = _callIfExists2.default

var _interpolate2 = _interopRequireDefault(require('./interpolate'))

exports.interpolate = _interpolate2.default

var _raf = _interopRequireWildcard(require('raf'))

exports.requestAnimationFrame = _raf.default
exports.cancelAnimationFrame = _raf.cancel

var _requestInterval2 = _interopRequireWildcard(require('./requestInterval'))

exports.requestInterval = _requestInterval2.default
exports.clearRequestInterval = _requestInterval2.clearRequestInterval

var _requestTimeout2 = _interopRequireWildcard(require('./requestTimeout'))

exports.requestTimeout = _requestTimeout2.default
exports.clearRequestTimeout = _requestTimeout2.clearRequestTimeout

var _strictShallowEqual2 = _interopRequireDefault(
  require('./strictShallowEqual')
)

exports.strictShallowEqual = _strictShallowEqual2.default

var _toKebabCase2 = _interopRequireWildcard(require('./toKebabCase'))

exports.toKebabCase = _toKebabCase2.default
exports.toKebabCaseTrimmed = _toKebabCase2.toKebabCaseTrimmed

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj
  } else {
    var newObj = {}
    if (obj != null) {
      for (var key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          var desc =
            Object.defineProperty && Object.getOwnPropertyDescriptor
              ? Object.getOwnPropertyDescriptor(obj, key)
              : {}
          if (desc.get || desc.set) {
            Object.defineProperty(newObj, key, desc)
          } else {
            newObj[key] = obj[key]
          }
        }
      }
    }
    newObj.default = obj
    return newObj
  }
}

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj}
}
