// @render-props/viewport/cjs/ViewportProvider.js
'use strict'

var _interopRequireDefault = require('@babel/runtime/helpers/interopRequireDefault')

exports.__esModule = true
exports.default = ViewportProvider

var _react = _interopRequireDefault(require('react'))

var _propTypes = _interopRequireDefault(require('prop-types'))

var _Viewport = _interopRequireDefault(require('./Viewport'))

var _ViewportContext = _interopRequireDefault(require('./ViewportContext'))

function ViewportProvider(props) {
  let prevContext = {}
  return (0, _Viewport.default)({
    children: function(context) {
      if (
        prevContext.scrollY !== context.scrollY ||
        prevContext.aspect !== context.aspect ||
        prevContext.scrollX !== context.scrollX
      ) {
        // ensures immutability for the provider value
        context = {
          inFullView: context.inFullView,
          inFullViewX: context.inFullViewX,
          inFullViewY: context.inFullViewY,
          inView: context.inView,
          inViewX: context.inViewX,
          inViewY: context.inViewY,
          scrollTo: context.scrollTo,
          scrollX: context.scrollX,
          scrollY: context.scrollY,
          direction: context.direction,
          distance: context.distance,
          orientation: context.orientation,
          screenOrientation: context.screenOrientation,
          height: context.height,
          width: context.width,
          aspect: context.aspect,
        }
        prevContext = context
      }

      return _react.default.createElement(
        _ViewportContext.default.Provider,
        {
          value: context,
        },
        props.children
      )
    },
  })
}

ViewportProvider.propTypes = {
  children: _propTypes.default.element.isRequired,
}
